### PR TITLE
Migrate Gulp scripts to Makefile, add latest tag for releases and update dependencies

### PR DESCRIPTION
**Pull request by maciaszczykm**
At the beginning I would like to apologize for the size of the it but once I started removing Gulp tasks I wanted to get rid of all of them. After few days of work it's done and most of the logic is moved to Makefile which is a lot shorter and easier to understand than the previous setup. There are still a lot of things that can be improved but at some point it has to be merged and then we can continue.

@floreks @shu-mutou and anyone else who wants to contribute please review and test it so we will avoid any regressions.

Progress:
- [x] add latest tag for releases
- [x] remove all tracks of Gulp and add Makefile
- [x] update documentation
- [x] update dependencies, Angular to v13 and Node to v16

Closes https://github.com/kubernetes/dashboard/issues/5417.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+ENABLE_INSECURE_LOGIN ?= false
+ENABLE_SKIP_LOGIN ?= false
+SYSTEM_BANNER ?= "Local test environment"
+SYSTEM_BANNER_SEVERITY ?= INFO
+PROD_BINARY = dist/amd64/dashboard
+SERVE_DIRECTORY = .tmp/serve
+SERVE_BINARY = .tmp/serve/dashboard
+SERVE_PID = .tmp/serve/dashboard.pid
+RELEASE_IMAGE = kubernetesui/dashboard
+RELEASE_VERSION = v2.3.1
```
Comment (shu-mutou) [inline Makefile:32]: v2.4.0

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,266 @@
+SHELL = /bin/bash
+GOOS ?= $(shell go env GOOS)
+GOARCH ?= $(shell go env GOARCH)
+GOPATH ?= $(shell go env GOPATH)
+ROOT_DIRECTORY := $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
+COVERAGE_DIRECTORY = $(ROOT_DIRECTORY)/coverage
+GO_COVERAGE_FILE = $(ROOT_DIRECTORY)/coverage/go.txt
+FSWATCH_BINARY := $(shell which fswatch)
+CODEGEN_VERSION := v0.22.3
```
Comment (shu-mutou) [inline Makefile:9]: Update other k8s modules by merging #6608 .

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+		--bind-address=127.0.0.1 \
+		--port=8080
+
+.PHONY: test-backend
+test-backend: ensure-go
+	go test $(MAIN_PACKAGE)/...
+
+.PHONY: test-frontend
+test-frontend:
+	npm run test
```
Comment (shu-mutou) [inline Makefile:139]: This results in infinite loop if we run `npm run test` from prompt.

Comment (shu-mutou) [reply Makefile:139]: `npx jest -c aio/jest.config.js`, right?

File: .npmrc [hunk 4]
```diff
@@ -1,17 +0,0 @@
...
-
-# Dashboard configuration
-kubeconfig_dir = ${HOME}/.kube
-kubeconfig = ${HOME}/.kube/config
-frontend_build_dir = dist/frontend/en
-dashboard_binary_path = dist/amd64/dashboard
-dashboard_locale_config = dist/amd64/locale_conf.json
-bind_address = 127.0.0.1
-sidecar_host = http://localhost:8000
-port = 8080
```
Comment (shu-mutou) [inline .npmrc:17]: I use these params. Could you enable to use them in new scripts?

Comment (maciaszczykm) [reply .npmrc:17]: See the top of the makefile. You can still specify them, i.e. "ENV_VAR=test make test".

File: Makefile (hunk 3 above)
Comment (maciaszczykm) [reply Makefile:139]: It was supposed to be `npm run test:frontend` but then I switched something. Thanks!

File: .npmrc (hunk 4 above)
Comment (maciaszczykm) [reply .npmrc:17]: Okay, I see now. Port and address are missing. I'll revert .npmrc file.

**Comment by floreks** [comment]
LGTM

@shu-mutou could you also take a look as you are using some run-in-docker configurations.

**Comment by shu-mutou** [comment]
Unfortunately nothing is going well for me, i.e. in development container.
I tried following versions of nodejs and golang.
* nodejs
  + 14.17.5: NG
  + 14.18.1: NG
  + 16.10.0: looks OK to run test and build, but not to run check.
* golang
  + 1.17: The k8s community had desided to use the version.  But `cosmtrek/air` seems to be built with golang 1.16 and makes `vendor` directory. Should we use it?

`make check-go` returns following errors:
```
user@d96b6fc90239:/go/src/github.com/kubernetes/dashboard$ make check-go
golangci-lint run -c .golangci.yml ./src/app/backend/...
ERRO Running error: context loading failed: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: go: inconsistent vendoring in /go/src/github.com/kubernetes/dashboard:
	github.com/PuerkitoBio/purell@v1.1.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/PuerkitoBio/urlesc@v0.0.0-20170810143723-de5bf2ad4578: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/emicklei/go-restful@v2.9.5+incompatible: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/go-openapi/jsonpointer@v0.19.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/go-openapi/jsonreference@v0.19.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/go-openapi/swag@v0.19.14: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/josharian/intern@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/mailru/easyjson@v0.7.6: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	golang.org/x/mod@v0.4.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	golang.org/x/tools@v0.1.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	k8s.io/code-generator@v0.22.3: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	k8s.io/gengo@v0.0.0-20201214224949-b6c5ce23f027: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt

	To ignore the vendor directory, use -mod=readonly or -mod=mod.
	To sync the vendor directory, run:
		go mod vendor 
make: *** [Makefile:175: check-go] Error 3

```

And `go mod vendor` returns follows:
```
user@d96b6fc90239:/go/src/github.com/kubernetes/dashboard$ go version
go version go1.17 linux/amd64
user@d96b6fc90239:/go/src/github.com/kubernetes/dashboard$ go mod vendor
go: finding module for package github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned/fake
go: finding module for package github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned
github.com/kubernetes/dashboard/src/app/backend/client imports
	github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned: cannot find module providing package github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned: module github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned: reading https://proxy.golang.org/github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned/@v/list: 410 Gone
	server response: not found: module github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned: no matching versions for query "latest"
github.com/kubernetes/dashboard/src/app/backend/plugin tested by
	github.com/kubernetes/dashboard/src/app/backend/plugin.test imports
	github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned/fake: cannot find module providing package github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned/fake: module github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned/fake: reading https://proxy.golang.org/github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned/fake/@v/list: 410 Gone
	server response: not found: module github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset/versioned/fake: no matching versions for query "latest"
```

Also, fail to install codegen.
```
go get -d k8s.io/code-generator@v0.22.3
chmod +x /go/pkg/mod/k8s.io/code-generator@v0.22.3/generate-groups.sh
./aio/scripts/verify-codegen.sh
Generating deepcopy funcs
Generating clientset for apis:v1alpha1 at github.com/kubernetes/dashboard/src/app/backend/plugin/client/clientset
Generating listers for apis:v1alpha1 at github.com/kubernetes/dashboard/src/app/backend/plugin/client/listers
Generating informers for apis:v1alpha1 at github.com/kubernetes/dashboard/src/app/backend/plugin/client/informers
mv: cannot stat './aio/scripts/../../aio/scripts/../github.com/kubernetes/dashboard/src/app/backend/plugin/apis/v1alpha1/zz_generated.deepcopy.go': No such file or directory
make: *** [Makefile:167: check-codegen] Error 1
```

Could you give advice for these?

File: .air.toml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+root = "."
+tmp_dir = "tmp"
```
Comment (shu-mutou) [inline .air.toml:2]: This creates `tmp` directory as not staged file, if `make` fails.

```suggestion
tmp_dir = ".tmp"
```

**Comment by shu-mutou** [comment]
However, recovering CI is the highest priority, so issues in development container can be postponed. 😉

**Comment by maciaszczykm** [comment]
@floreks @shu-mutou I've added few updates:
- fixed `make check` command (codegen check was not passing because I've removed `go mod tidy` by accident from script that installs code-generator)
- updated multiple dependencies including Angular to v13
- switched to Node v16
Please take another look. I'd really love to finally stop solving conflicts on this one if everything works as before :)

**Comment by floreks** [comment]
@shu-mutou could you recheck on your side after the changes. Everything seems to be working for me just fine.

**Comment by shu-mutou** [comment]
Checks, tests and build are working! 😄 
But translations are not shown. 😢 

For development container, npm variables are not passed to dashboard, but I'd like to create new PR for fitting their treatments to new processes.

**Comment by maciaszczykm** [comment]
>But translations are not shown. 😢

@shu-mutou Which start method are you using? I think translations should be visible in the container but perhaps there is some issue with local dev scripts.

**Comment by shu-mutou** [comment]
I did `make build` and run `dist/amd64/dashboard --kubeconfig ......`.
Only English was displayed last week, but now every translations are displayed. 🤔 

Anyway, this goes well! Thanks for great job!!
/lgtm

**Pull request merged**